CredentialsManager/Models/Credentials.cs

```diff
@@ -0,0 +1,100 @@
+﻿using System;
+
+namespace CredentialsManager.Models
+{
+  public class Credentials : IEquatable<Credentials>, IComparable<Credentials>, IComparable
+  {
+    public string UserName { get; }
+    public string UserPassword { get; }
+
+    public Credentials(string userName, string userPassword)
+    {
+      if (string.IsNullOrWhiteSpace(userName))
+      {
+        throw new ArgumentException("Value can not be null or empty", nameof(userName));
+      }
+
+      if (string.IsNullOrWhiteSpace(userPassword))
+      {
+        throw new ArgumentException("Value can not be null or empty", nameof(userName));
+      }
+      
+      UserName = userName;
+      UserPassword = userPassword;
+    }
+
+    public static bool TryParse(string line, string delimiter, out Credentials credentials)
+    {
+      credentials = default;
+      const int validRowLength = 2;
+
+      if (string.IsNullOrWhiteSpace(line)) return false;
+
+      var row = line.Split(delimiter);
+      if (row.Length != validRowLength) return false;
+
+      var userName = row[0].Trim();
+      var userPassword = row[1].Trim();
+
+      if (string.IsNullOrWhiteSpace(userName) || string.IsNullOrWhiteSpace(userPassword))
+      {
+        return false;
+      }
+
+      credentials = new Credentials(userName, userPassword);
+
+      return true;
+    }
+
+    public override string ToString()
+    {
+      return $"{UserName},{UserPassword}";
+    }
+
+    public bool Equals(Credentials? other)
+    {
+      if (other is null) return false;
+      if (ReferenceEquals(this, other)) return true;
+
+      return UserName.Equals(other.UserName, StringComparison.OrdinalIgnoreCase)
+             && UserPassword.Equals(other.UserPassword, StringComparison.Ordinal);
+    }
+
+    public override bool Equals(object? obj)
+    {
+      if (obj is null) return false;
+      if (ReferenceEquals(this, obj)) return true;
+
+      return obj.GetType() == GetType() && Equals((Credentials)obj);
+    }
+
+    public override int GetHashCode()
+    {
+      return HashCode.Combine(UserName.ToLowerInvariant(), UserPassword);
```

Comment:
If your `Equals` uses `StringComparison.Ordinal` and your `HashCode` uses `ToLowerInvariant` you can get subtle bugs with unicode string expansion since ordinal is not the same thing as invariant.

Reply:
I managed to find a proper solution for this (at least I think I do :) )
- I used `OrdinalIgnoreCase` just for `UserName` as the pass should not change case
- I created a `HashCode` object inside `GetHasCode` method, to which I added `UserName` with `OrdinalIgnoreCase`, added the password normally then I convert it and return as the actual hash code int value.

I updated it in this commit: [b7a6dd3](https://github.com/csinn/CSharp-From-Zero-To-Hero-v2/pull/43/commits/b7a6dd3db544d5f8ba4b35d76dc1f478e7268391)